src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	public class UnconditionalSuppressMessageAttributeState
+	{
+		private readonly AssemblyDefinition _assembly;
+		private readonly Dictionary<IMetadataTokenProvider, Dictionary<string, SuppressMessageInfo>> _localSuppressionsByMdToken;
+
+		private bool HasLocalSuppressions {
+			get {
+				return _localSuppressionsByMdToken.Count != 0;
+			}
+		}
+
+		public UnconditionalSuppressMessageAttributeState (AssemblyDefinition assembly)
+		{
+			_assembly = assembly;
+			_localSuppressionsByMdToken = new Dictionary<IMetadataTokenProvider, Dictionary<string, SuppressMessageInfo>> ();
+		}
+
+		public void AddLocalSuppression (CustomAttribute ca, IMetadataTokenProvider mdTokenProvider)
+		{
+			SuppressMessageInfo info;
+			if (!TryDecodeSuppressMessageAttributeData (ca, out info)) {
+				return;
+			}
+
+			if (!_localSuppressionsByMdToken.TryGetValue (mdTokenProvider, out var suppressions)) {
+				suppressions = new Dictionary<string, SuppressMessageInfo> ();
+				_localSuppressionsByMdToken.Add (mdTokenProvider, suppressions);
+			}
+
+			suppressions.Add (info.Id, info);
```

Comment:
Is there a guarantee that there will be only one "info" for each ID? Otherwise this is going to throw at runtime. I think this should be either a Dictionary<string, List<>> or just a flat list which we search through each time.

Reply:
There's not. I've changed this to `TryAdd`, since I see no scenario in which we should store two "info" for the same (md token provider, ID) tuple.

Reply:
Changed it once again since `TryAdd` is only supported in .NET Core. The behavior now is to keep the last `info` seen for these (provider, id) tuples.